cpp/src/arrow/filesystem/filesystem.h

```diff
@@ -0,0 +1,181 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <chrono>
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "arrow/status.h"
+#include "arrow/util/compression.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+namespace io {
+
+class InputStream;
+class OutputStream;
+class RandomAccessFile;
+
+}  // namespace io
+
+namespace fs {
+
+// A system clock time point expressed as a 64-bit (or more) number of
+// nanoseconds since the epoch.
+using TimePoint =
+    std::chrono::time_point<std::chrono::system_clock, std::chrono::nanoseconds>;
+
+/// \brief EXPERIMENTAL: FileSystem entry type
+enum class ARROW_EXPORT FileType {
+  // Target does not exist
+  NonExistent,
+  // Target exists but its type is unknown (could be a special file such
+  // as a Unix socket or character device, or Windows NUL / CON / ...)
+  Unknown,
+  // Target is a regular file
+  File,
+  // Target is a directory
+  Directory
+};
+
+ARROW_EXPORT std::string ToString(FileType);
+
+static const int64_t kNoSize = -1;
+static const TimePoint kNoTime = TimePoint(TimePoint::duration(-1));
+
+/// \brief EXPERIMENTAL: FileSystem entry stats
+struct ARROW_EXPORT FileStats {
+  FileStats() = default;
+  FileStats(FileStats&&) = default;
+  FileStats& operator=(FileStats&&) = default;
+  FileStats(const FileStats&) = default;
+  FileStats& operator=(const FileStats&) = default;
+
+  // The file type.
+  FileType type() const { return type_; }
+  void SetType(FileType type) { type_ = type; }
+
+  // The full file path in the filesystem.
+  std::string path() const { return path_; }
+  void SetPath(const std::string& path) { path_ = path; }
```

Comment:
These should probably be `set_type` as they are low-cost methods (that's the deciding factor for CamelCase or not in the Google Styleguide)

Reply:
Will do :-)